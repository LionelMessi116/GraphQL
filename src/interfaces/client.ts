import Observable from 'zen-observable-ts';

import { ICache } from './cache';
import { IExchangeResult } from './exchange';
import { IQuery } from './query';

export interface IClient {
  cache: ICache;
  executeQuery$(
    queryObject: IQuery,
    skipCache: boolean
  ): Observable<IExchangeResult>;
  executeQuery(
    queryObject: IQuery,
    skipCache: boolean
  ): Promise<IExchangeResult>;
  executeMutation$(mutationObject: IQuery): Observable<IExchangeResult['data']>;
  executeMutation(mutationObject: IQuery): Promise<IExchangeResult['data']>;
  refreshAllFromCache(): void;
  subscribe(
    callback: (changedTypes: string[], reponse: object) => void
  ): () => void;
}
