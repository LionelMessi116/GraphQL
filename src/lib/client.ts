import {
  filter,
  makeSubject,
  pipe,
  publish,
  share,
  Source,
  Subject,
  subscribe,
  take,
} from 'wonka';

import * as uuid from 'uuid';
import {
  cacheExchange,
  dedupeExchange,
  fetchExchange,
  subscriptionExchange,
} from '../exchanges';
import { CombinedError, hashString } from '../lib';
import {
  Client,
  ClientInstance,
  ClientOptions,
  CreateClientInstanceOpts,
  Exchange,
  ExchangeResult,
  Mutation,
  Operation,
  OperationType,
  Query,
  Subscription,
  SubscriptionStreamUpdate,
} from '../types';

type SubscriptionMap = Map<string, () => void>;

const defaultExchanges = [dedupeExchange, cacheExchange, fetchExchange];

/** Function for creating an application wide [Client]{@link Client}. */
export const createClient = (opts: ClientOptions): Client => {
  /** Cache of all instance subscriptions */
  const subscriptions = new Map<string, SubscriptionMap>();
  const querySubscriptions = new Map<string, SubscriptionMap>();

  /** Main subject for emitting operations */
  const subject = makeSubject<Operation>();

  const exchanges =
    opts.exchanges !== undefined ? opts.exchanges : defaultExchanges;

  /** Main subject stream for listening to operation responses */
  const subject$ = pipe(
    subject[0],
    pipeExchanges(exchanges, subject),
    share
  );

  publish(subject$);

  const fetchOptions =
    typeof opts.fetchOptions === 'function'
      ? opts.fetchOptions()
      : opts.fetchOptions;

  /** Convert a query to an operation type */
  const createOperation = (
    type: OperationType,
    query: Query,
    options: object = {}
  ) => ({
    ...query,
    key: hashString(JSON.stringify(query)),
    operationName: type,
    context: {
      url: opts.url,
      fetchOptions,
      forwardSubscription: opts.forwardSubscription,
      ...options,
    },
  });

  const executeOperation = (operation: Operation) => subject[1](operation);

  /** Function to manage subscriptions on a Consumer by Consumer basis */
  const createInstance = (
    instanceOpts: CreateClientInstanceOpts
  ): ClientInstance => {
    const id = uuid.v4();

    const getSubscriptions = () =>
      subscriptions.has(id) ? subscriptions.get(id) : (new Map() as SubscriptionMap);

    const updateSubscriptions = (subs: SubscriptionMap) =>
      subscriptions.set(id, subs);

    const executeQuery = (query: Query, force?: boolean) => {
      const operation = createOperation(OperationType.Query, query, { force });

      const outboundKey = `${operation.key}-out`;
      const inboundKey = `${operation.key}-in`;

      const instanceSubscriptions = getSubscriptions();
      if (!instanceSubscriptions.has(outboundKey)) {
        /** Notify component when fetching query is occurring */
        const [outboundSub] = pipe(
          subject[0],
          filter(op => op.key === operation.key),
          subscribe(() => instanceOpts.onChange({ fetching: true }))
        );

        instanceSubscriptions.set(outboundKey, outboundSub);
      }

      if (!instanceSubscriptions.has(inboundKey)) {
        /** Notify component when query operation has returned */
        const [inboundSub] = pipe(
          subject$,
          filter(result => result.operation.key === operation.key),
          subscribe(response => {
            instanceOpts.onChange({
              data: response.data,
              error: response.error,
              fetching: false,
            });
          })
        );

        instanceSubscriptions.set(inboundKey, inboundSub);
      }

      updateSubscriptions(instanceSubscriptions);
      executeOperation(operation);
    };

    const executeMutation = (query: Mutation) => {
      const operation = createOperation(OperationType.Mutation, query);
      executeOperation(operation);
    };

    const executeSubscription = (query: Subscription) => {
      const operation = createOperation(OperationType.Subscription, query);
      const subscriptionSubject = makeSubject<Operation>();

      querySubscriptions.set(query.query, subscriptionSubject);

      /** Notify component when subscription operation has been given a value */
      subscriptionSubject
        .pipe(
          pipeExchanges(
            opts.exchanges !== undefined ? opts.exchanges : defaultExchanges,
            subject
          )
        )
        .subscribe((response: SubscriptionStreamUpdate) => {
          instanceOpts.onSubscriptionUpdate({
            data: response.data.data,
            error: Array.isArray(response.data.errors)
              ? new CombinedError({
                  graphQLErrors: response.data.errors,
                  response,
                })
              : response.error,
          });
        });

      executeOperation(operation);
    };

    const executeUnsubscribeSubscription = (query: Subscription) => {
      const operation = createOperation(OperationType.Subscription, query, {
        unsubscribe: true,
      });

      // remove cached rx subscription
      const activeSubscriptionSubject = querySubscriptions.get(query.query);

      if (activeSubscriptionSubject !== undefined) {
        querySubscriptions.delete(query.query);
        executeOperation(operation);

        // This removes the previous existence of a subscription from being part of the subject
        // without this, any unsubscribe -> subscribe will have an n^2 effect.
        activeSubscriptionSubject.unsubscribe();
      }
    };

    const unsubscribe = () => {
      const subs = getRxSubscriptions();
      if (subs === undefined) {
        return;
      }

      [...subs.values()].forEach(unsub => unsub());
    };

    return {
      executeQuery,
      executeMutation,
      executeSubscription,
      executeUnsubscribeSubscription,
      unsubscribe,
    };
  };

  return {
    createInstance,
  };
};

/** Create pipe of exchanges */
const pipeExchanges = (exchanges: Exchange[], subject: Subject<Operation>) => (
  operation$: Source<Operation>
): Source<ExchangeResult> => {
  /** Recursively pipe to each exchange */
  const callExchanges = (value: Source<Operation>, index: number = 0) => {
    if (exchanges.length < index) {
      return value;
    }

    const currentExchange = exchanges[index];
    return currentExchange({
      forward: val => callExchanges(val, index + 1),
      subject,
    })(value);
  };

  return callExchanges(operation$, 0);
};
