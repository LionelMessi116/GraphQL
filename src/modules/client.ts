import Observable from 'zen-observable-ts';

import {
  ICache,
  IClientOptions,
  IExchange,
  IExchangeResult,
  IQuery,
} from '../interfaces/index';

import { cacheExchange } from './cache-exchange';
import { dedupExchange } from './dedup-exchange';
import { defaultCache } from './default-cache';
import { hashString } from './hash';
import { httpExchange } from './http-exchange';

export default class Client {
  url: string;
  store: object; // Internal store
  subscriptions: object; // Map of subscribed Connect components
  subscriptionSize: number; // Used to generate IDs for subscriptions
  cache: ICache; // Cache object
  exchange: IExchange; // Exchange to communicate with GraphQL APIs
  fetchOptions: RequestInit | (() => RequestInit); // Options for fetch call

  constructor(opts?: IClientOptions) {
    if (!opts) {
      throw new Error('Please provide configuration object');
    } else if (!opts.url) {
      throw new Error('Please provide a URL for your GraphQL API');
    }

    this.url = opts.url;
    this.store = opts.initialCache || {};
    this.subscriptions = {};
    this.subscriptionSize = 0;
    this.cache = opts.cache || defaultCache(this.store);
    this.exchange = cacheExchange(this.cache, dedupExchange(httpExchange()));
    this.fetchOptions = opts.fetchOptions || {};

    // Bind methods
    this.executeQuery = this.executeQuery.bind(this);
    this.executeMutation = this.executeMutation.bind(this);
    this.updateSubscribers = this.updateSubscribers.bind(this);
    this.refreshAllFromCache = this.refreshAllFromCache.bind(this);
  }

  updateSubscribers(typenames, changes) {
    // On mutation, call subscribed callbacks with eligible typenames
    for (const sub in this.subscriptions) {
      if (this.subscriptions.hasOwnProperty(sub)) {
        this.subscriptions[sub](typenames, changes);
      }
    }
  }

  subscribe(
    callback: (changedTypes: string[], response: object) => void
  ): () => void {
    // Create an identifier, add callback to subscriptions
    const id = this.subscriptionSize++;
    this.subscriptions[id] = callback;

    // Return unsubscription function
    return () => {
      delete this.subscriptions[id];
    };
  }

  refreshAllFromCache() {
    // On mutation, call subscribed callbacks with eligible typenames
    return new Promise(resolve => {
      for (const sub in this.subscriptions) {
        if (this.subscriptions.hasOwnProperty(sub)) {
          this.subscriptions[sub](null, null, true);
        }
      }
      resolve();
    });
  }

  makeContext({ skipCache }: { skipCache?: boolean }): Record<string, any> {
    return {
      fetchOptions:
        typeof this.fetchOptions === 'function'
          ? this.fetchOptions()
          : this.fetchOptions,
      skipCache: !!skipCache,
      url: this.url,
    };
  }

  executeQuery$(
    queryObject: IQuery,
    skipCache: boolean
  ): Observable<IExchangeResult> {
    // Create hash key for unique query/variables
    const { query, variables } = queryObject;
    const key = hashString(JSON.stringify({ query, variables }));
    const operation = {
      context: this.makeContext({ skipCache }),
      key,
      operationName: 'query',
      query,
      variables,
    };

    return this.exchange(operation);
  }

  executeQuery(
    queryObject: IQuery,
    skipCache: boolean
  ): Promise<IExchangeResult> {
    return new Promise<IExchangeResult>((resolve, reject) => {
      this.executeQuery$(queryObject, skipCache).subscribe({
        error: reject,
        next: resolve,
      });
    });
  }

  executeMutation$(
    mutationObject: IQuery
  ): Observable<IExchangeResult['data']> {
    // Create hash key for unique query/variables
    const { query, variables } = mutationObject;
    const key = hashString(JSON.stringify({ query, variables }));

    const operation = {
      context: this.makeContext({}),
      key,
      operationName: 'mutation',
      query,
      variables,
    };

    return this.exchange(operation).map((response: IExchangeResult) => {
      // Notify subscribed Connect wrappers
      this.updateSubscribers(response.typeNames, response);
      // Resolve result
      return response.data;
    });
  }

  executeMutation(mutationObject: IQuery): Promise<IExchangeResult['data']> {
    return new Promise<IExchangeResult>((resolve, reject) => {
      this.executeMutation$(mutationObject).subscribe({
        error: reject,
        next: resolve,
      });
    });
  }
}
