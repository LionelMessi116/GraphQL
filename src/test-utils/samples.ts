import {
  ExchangeResult,
  ExecutionResult,
  Mutation,
  Operation,
  Query,
  Subscription,
} from '../types';

const context = {
  fetchOptions: {
    method: 'POST',
  },
  url: 'http://localhost:3000/graphql',
};

export const queryGql: Query = {
  query: `query getUser($name: String){
    user(name: $name) {
      id
      firstName
      lastName
    }
  }`,
  variables: {
    name: 'Clara',
  },
};

export const mutationGql: Mutation = {
  query: `mutation AddUser($name: String){
    addUser(name: $name) {
      name
    }
  }`,
  variables: {
    name: 'Clara',
  },
};

export const subscriptionGql: Subscription = {
  query: `subscription subscribeToUser($user: String){
    user(user: $user) {
      name
    }
  }
  `,
  variables: {
    user: 'colin',
  },
};

export const queryOperation: Operation = {
  key: '2',
  operationName: 'query',
  context,
  ...queryGql,
};

export const mutationOperation: Operation = {
  key: JSON.stringify(mutationGql),
  operationName: 'mutation',
  context,
  ...mutationGql,
};

export const subscriptionOperation: Operation = {
  key: JSON.stringify(subscriptionGql),
  operationName: 'subscription',
  context,
  ...subscriptionGql,
};

export const queryResponse: ExchangeResult = {
  operation: queryOperation,
  data: {
    user: {
      name: 'Clive',
    },
  },
};

export const mutationResponse: ExchangeResult = {
  operation: mutationOperation,
  data: {},
};

export const subscriptionResult: ExecutionResult = {
  data: {}
};
