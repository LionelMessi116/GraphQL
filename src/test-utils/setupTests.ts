import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Setup enzyme adapter
configure({ adapter: new Adapter() });

// Setup globals
const g = global as any;

g.AbortController = undefined;
g.fetch = jest.fn();

process.on('unhandledRejection', error => {
  throw error;
});
