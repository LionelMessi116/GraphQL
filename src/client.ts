import { filter, makeSubject, onStart, pipe, share, Source, take } from 'wonka';
import { composeExchanges, defaultExchanges } from './exchanges';
import { hashString } from './lib';

import {
  Exchange,
  GraphqlMutation,
  GraphqlQuery,
  GraphQLRequest,
  GraphqlSubscription,
  Operation,
  OperationContext,
  OperationResult,
  OperationType,
} from './types';

/** Options for configuring the URQL [client]{@link Client}. */
export interface ClientOptions {
  /** Target endpoint URL such as `https://my-target:8080/graphql`. */
  url: string;
  /** Any additional options to pass to fetch. */
  fetchOptions?: RequestInit | (() => RequestInit);
  /** An ordered array of Exchanges. */
  exchanges?: Exchange[];
}

interface ActiveResultSources {
  [operationKey: string]: Source<OperationResult>;
}

export const createClient = (opts: ClientOptions) => new Client(opts);

/** The URQL application-wide client library. Each execute method starts a GraphQL request and returns a stream of results. */
export class Client implements ClientOptions {
  // These are variables derived from ClientOptions
  url: string;
  fetchOptions: RequestInit;
  exchanges: Exchange[];

  // These are internals to be used to keep track of operations
  dispatchOperation: (operation: Operation) => void;
  operations$: Source<Operation>;
  results$: Source<OperationResult>;
  activeResultSources = Object.create(null) as ActiveResultSources;

  constructor(opts: ClientOptions) {
    this.url = opts.url;

    this.fetchOptions =
      typeof opts.fetchOptions === 'function'
        ? opts.fetchOptions()
        : opts.fetchOptions || {};

    this.exchanges =
      opts.exchanges !== undefined ? opts.exchanges : defaultExchanges;

    // This subject forms the input of operations; executeOperation may be
    // called to dispatch a new operation on the subject
    const [operations$, nextOperation] = makeSubject<Operation>();
    this.operations$ = operations$;
    this.dispatchOperation = nextOperation;

    // All operations run through the exchanges in a pipeline-like fashion
    // and this observable then combines all their results
    this.results$ = share(
      composeExchanges(this, this.exchanges)(this.operations$)
    );
  }

  private createOperationContext = (
    opts?: Partial<OperationContext>
  ): OperationContext => ({
    url: this.url,
    fetchOptions: this.fetchOptions,
    ...opts,
  });

  private createRequestOperation = (
    type: OperationType,
    query: GraphQLRequest,
    opts?: Partial<OperationContext>
  ): Operation => ({
    ...query,
    key: hashString(JSON.stringify(query)),
    operationName: type,
    context: this.createOperationContext(opts),
  });

  /** Deletes an active operation's result observable and sends a teardown signal through the exchange pipeline */
  // private teardownOperation(operation: Operation) {
  //   delete this.activeResultSources[operation.key];
  //   this.dispatchOperation({ ...operation, operationName: 'teardown' });
  // }

  /** Executes an Operation by sending it through the exchange pipeline It returns an observable that emits all related exchange results and keeps track of this observable's subscribers. A teardown signal will be emitted when no subscribers are listening anymore. */
  executeRequestOperation(operation: Operation): Source<OperationResult> {
    const { key, operationName } = operation;

    const operationResults$ = pipe(
      this.results$,
      filter(res => res.operation.key === key)
    );

    if (operationName === 'mutation') {
      // A mutation is always limited to just a single result and is never shared
      return pipe(
        operationResults$,
        onStart<OperationResult>(() => this.dispatchOperation(operation)),
        take(1)
      );
    }

    if (key in this.activeResultSources) {
      // Reuse active result observable when it's available
      return this.activeResultSources[key];
    }

    return (this.activeResultSources[key] = pipe(
      operationResults$,
      onStart<OperationResult>(() => this.dispatchOperation(operation)),
      // onEnd<OperationResult>(() => this.teardownOperation(operation)),
      share
    ));
  }

  reexecuteOperation = (operation: Operation) => {
    // Reexecute operation only if any subscribers are still subscribed to the
    // operation's exchange results
    if (this.activeResultSources[operation.key] !== undefined) {
      this.dispatchOperation(operation);
    }
  };

  executeQuery = (
    query: GraphqlQuery,
    opts?: Partial<OperationContext>
  ): Source<OperationResult> => {
    const operation = this.createRequestOperation('query', query, opts);
    return this.executeRequestOperation(operation);
  };

  executeSubscription = (
    query: GraphqlSubscription,
    opts?: Partial<OperationContext>
  ): Source<OperationResult> => {
    const operation = this.createRequestOperation('subscription', query, opts);
    return this.executeRequestOperation(operation);
  };

  executeMutation = (
    query: GraphqlMutation,
    opts?: Partial<OperationContext>
  ): Source<OperationResult> => {
    const operation = this.createRequestOperation('mutation', query, opts);
    return this.executeRequestOperation(operation);
  };
}
