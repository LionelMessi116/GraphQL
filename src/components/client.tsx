import { Component, ReactNode } from 'react';
import { CombinedError } from '../lib';
import {
  ChildArgs,
  Client,
  ClientInstance,
  Mutation,
  Query,
  StreamUpdate,
} from '../types';

export interface ClientState<MutationDeclarations> {
  client: ClientInstance;
  fetching: boolean;
  error?: Error | CombinedError | CombinedError[];
  data?: any | any[];
  mutations?: { [type in keyof MutationDeclarations]: (vals: any) => void };
}

export interface ClientProps<MutationDeclarations> {
  client: Client;
  children: (obj: ChildArgs<MutationDeclarations>) => ReactNode;
  query?: Query;
  mutations?: { [type in keyof MutationDeclarations]: Mutation };
}

/** Component responsible for implementing the [Client]{@link Client} utility in React. */
export class UrqlClient<MutationDeclarations> extends Component<
  ClientProps<MutationDeclarations>,
  ClientState<MutationDeclarations>
> {
  constructor(props: ClientProps<MutationDeclarations>) {
    super(props);

    this.state = {
      client: props.client.createInstance({
        onChange: result => this.onStreamUpdate(result),
      }),
      data: undefined,
      error: undefined,
      fetching: true,
      mutations: this.getMutatorFunctions(),
    };
  }

  public componentDidMount() {
    if (this.props.query !== undefined) {
      this.state.client.executeQuery(this.props.query);
    }
  }

  public componentDidUpdate(prevProps: ClientProps<MutationDeclarations>) {
    if (
      JSON.stringify(prevProps.mutations) !==
      JSON.stringify(this.props.mutations)
    ) {
      this.setState({ mutations: this.getMutatorFunctions() });
    }

    if (
      this.props.query !== undefined &&
      JSON.stringify(prevProps.query) !== JSON.stringify(this.props.query)
    ) {
      this.state.client.executeQuery(this.props.query);
    }
  }

  public componentWillUnmount() {
    this.state.client.unsubscribe();
  }

  public render() {
    if (typeof this.props.children !== 'function') {
      return null;
    }

    return this.props.children({
      fetching: this.state.fetching,
      error: this.state.error,
      data: this.state.data,
      mutations: this.state.mutations,
      refetch: (noCache: boolean = false) =>
        this.props.query !== undefined
          ? this.state.client.executeQuery(this.props.query, noCache)
          : undefined,
    });
  }

  private getMutatorFunctions() {
    if (this.props.mutations === undefined) {
      return undefined;
    }

    const mutations = Object.keys(this.props.mutations).reduce(
      (prev, key) => ({
        ...prev,
        [key]: (variables: any) =>
          this.state.client.executeMutation({
            // @ts-ignore -  mutation key definitely exists
            ...this.props.mutations[key],
            variables,
          }),
      }),
      {}
    ) as ClientState<MutationDeclarations>['mutations'];

    return mutations;
  }

  private onStreamUpdate(updated: StreamUpdate) {
    this.setState({
      fetching: updated.fetching,
      error: updated.error,
      data: updated.data,
    });
  }
}
