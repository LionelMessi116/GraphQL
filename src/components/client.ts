import { Component, ReactNode } from 'react';
import { empty, filter, makeSubject, map, pipe, Source, subscribe, switchAll, toPromise } from 'wonka';

import { Client } from '../lib/client';
import { CombinedError } from '../lib/error';
import { ConnectProps } from './connect';

import {
  ChildArgs,
  ExchangeResult,
  Mutation,
  Query,
  Subscription
} from '../types';

export type Mutations<MutationDeclarations> = {
  [type in keyof MutationDeclarations]: Mutation
};

export type MutationFns<MutationDeclarations> = {
  [type in keyof MutationDeclarations]: (variables?: object) => Promise<ExchangeResult>
};

export interface ClientState<MutationDeclarations> {
  loaded: boolean;
  error?: CombinedError;
  data?: any;
  mutations?: MutationFns<MutationDeclarations>;
}

export interface ClientProps<MutationDeclarations> extends ConnectProps<MutationDeclarations> {
  client: Client;
}

/** Component responsible for implementing the [Client]{@link Client} utility in React. */
export class UrqlClient<MutationDeclarations> extends Component<
  ClientProps<MutationDeclarations>,
  ClientState<MutationDeclarations>
> {
  unsubscribe: () => void;
  nextQueryProp: (query?: Query) => void;
  nextSubscriptionProp: (subscription?: Subscription) => void;
  mutations = getMutationFns(this.props.client, this.props.mutations);

  state = {
    data: undefined,
    error: undefined,
    loaded: false,
  };

  constructor(props) {
    super(props);

    if (process.env.NODE_ENV !== 'production' && props.subscription && !props.updateSubscription) {
      throw new Error(
        'You instantiated an Urql Client component with a subscription but forgot an updateSubscription prop. updateSubscription callbacks are required to work with subscriptions.'
      );
    }

    const [queryProp$, nextQueryProp] = makeSubject<void | Query>();
    const [subscriptionProp$, nextSubscriptionProp] = makeSubject<void | Subscription>();
    this.nextQueryProp = nextQueryProp;
    this.nextSubscriptionProp = nextSubscriptionProp;

    const queryResults$ = makeQueryResults$(props.client, queryProp$);
    const subscriptionResults$ = makeSubscriptionResults$(props.client, subscriptionProp$);

    const [queryTeardown] = pipe(queryResults$, subscribe(this.onQueryUpdate));
    const [subscriptionTeardown] = pipe(queryResults$, subscribe(this.onSubscriptionUpdate));

    this.unsubscribe = () => {
      queryTeardown();
      subscriptionTeardown();
    };
  }

  private onQueryUpdate(result: ExchangeResult) {
    this.setState({
      loaded: true,
      error: result.error,
      data: result.data
    });
  }

  private onSubscriptionUpdate({ data, error }: ExchangeResult) {
    if (process.env.NODE_ENV !== 'production' && error) {
      throw error;
    }

    this.setState({
      loaded: true,
      error,
      data: this.props.updateSubscription
        ? this.props.updateSubscription(this.state.data, data)
        : this.state.data,
    });
  }

  componentDidMount() {
    this.nextQueryProp(this.props.query);
    this.nextSubscriptionProp(this.props.subscription);
  }

  componentDidUpdate(prevProps: ClientProps<MutationDeclarations>) {
    this.nextQueryProp(this.props.query);
    this.nextSubscriptionProp(this.props.subscription);

    if (prevProps.mutations !== this.props.mutations) {
      this.mutations = getMutationFns(this.props.client, this.props.mutations);
    }
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  render() {
    if (typeof this.props.children !== 'function') {
      return null;
    }

    return this.props.children({
      loaded: this.state.loaded,
      error: this.state.error,
      data: this.state.data,
      mutations: this.mutations || undefined
    });
  }
}

const makeQueryResults$ = (
  client: Client,
  queryProp$: Source<void | Query>
): Source<ExchangeResult> => {
  const noopQuery = (empty as Source<ExchangeResult>);
  let lastResults$: void | Source<ExchangeResult>;

  return pipe(
    queryProp$,
    map(query => {
      return query === undefined
        ? noopQuery
        : client.executeQuery(query);
    }),
    filter(x => {
      const isDistinct = x !== lastResults$;
      if (isDistinct) { lastResults$ = x; }
      return isDistinct;
    }),
    switchAll
  );
};

const makeSubscriptionResults$ = (
  client: Client,
  subscriptionProp$: Source<void | Subscription>
): Source<ExchangeResult> => {
  const noopSubscription = (empty as Source<ExchangeResult>);
  let lastResults$: void | Source<ExchangeResult>;

  return pipe(
    subscriptionProp$,
    map(query => {
      return query === undefined
        ? noopSubscription
        : client.executeSubscription(query);
    }),
    filter(x => {
      const isDistinct = x !== lastResults$;
      if (isDistinct) { lastResults$ = x; }
      return isDistinct;
    }),
    switchAll
  );
};

const getMutationFns = <T>(
  client: Client,
  mutations?: Mutations<T>
): void | MutationFns<T> => {
  if (mutations === undefined) {
    return undefined;
  }

  return Object.keys(mutations).reduce(
    (acc, key) => {
      acc[key] = (variables?: object) => pipe(
        client.executeMutation({
          ...mutations[key],
          variables,
        }),
        toPromise
      );

      return acc;
    },
    {} as MutationFns<T>
  );
}
