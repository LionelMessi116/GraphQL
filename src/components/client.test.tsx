import React from 'react';
import { shallow, mount } from 'enzyme';
import { UrqlClient } from './client';
import { queryGql as query, mutationGql } from '../test-utils';

const children = jest.fn().mockReturnValue(<h1>This is a child component</h1>);
const clientInstance = {
  executeQuery: jest.fn(),
  executeMutation: jest.fn(),
  unsubscribe: jest.fn(),
};
const client = { createInstance: jest.fn().mockReturnValue(clientInstance) };
const mutations = {
  addUser: mutationGql,
  removeUser: { ...mutationGql, query: `mutation removeUser() {}` },
};

beforeEach(() => {
  children.mockClear();
  client.createInstance.mockClear();
  clientInstance.executeQuery.mockClear();
  clientInstance.executeMutation.mockClear();
  clientInstance.unsubscribe.mockClear();
});

it('passes snapshot', () => {
  const wrapper = shallow(
    // @ts-ignore
    <UrqlClient client={client} query={query} mutations={mutations}>
      {children}
    </UrqlClient>
  );

  expect(wrapper).toMatchSnapshot();
});

describe('on creation', () => {
  it('calls createInstance', () => {
    shallow(
      // @ts-ignore
      <UrqlClient client={client} query={query} mutations={mutations}>
        {children}
      </UrqlClient>
    );

    expect(client.createInstance).toBeCalled();
  });

  describe('defaults', () => {
    it('passes snapshot', () => {
      shallow(
        // @ts-ignore
        <UrqlClient client={client} query={query} mutations={mutations}>
          {children}
        </UrqlClient>
      );

      expect(children.mock.calls).toMatchSnapshot();
    });

    it('fetching = true', () => {
      shallow(
        // @ts-ignore
        <UrqlClient client={client} query={query} mutations={mutations}>
          {children}
        </UrqlClient>
      );

      expect(children).toBeCalledWith(
        expect.objectContaining({ fetching: true })
      );
    });

    it('error = undefined', () => {
      shallow(
        // @ts-ignore
        <UrqlClient client={client} query={query} mutations={mutations}>
          {children}
        </UrqlClient>
      );

      expect(children).toBeCalledWith(
        expect.objectContaining({ error: undefined })
      );
    });

    it('data = undefined', () => {
      shallow(
        // @ts-ignore
        <UrqlClient client={client} query={query} mutations={mutations}>
          {children}
        </UrqlClient>
      );

      expect(children).toBeCalledWith(
        expect.objectContaining({ data: undefined })
      );
    });

    it('mutations = <object>', () => {
      shallow(
        // @ts-ignore
        <UrqlClient client={client} query={query} mutations={mutations}>
          {children}
        </UrqlClient>
      );

      expect(typeof children.mock.calls[0][0].mutations).toBe('object');
    });
  });
});

describe('componentDidMount', () => {
  it('calls executeQuery with query object', () => {
    const wrapper = shallow(
      // @ts-ignore
      <UrqlClient client={client} query={query}>
        {children}
      </UrqlClient>
    );

    wrapper.instance().componentDidMount();
    expect(clientInstance.executeQuery).toBeCalledWith(query);
  });
});

describe('componentDidUpdate', () => {
  it('query is refetched', () => {
    const wrapper = mount(
      // @ts-ignore
      <UrqlClient client={client} query={query}>
        {children}
      </UrqlClient>
    );

    const newQuery = { ...query, variables: { age: 40 } };

    children.mockClear();
    wrapper.setProps({ query: newQuery });
    expect(clientInstance.executeQuery).toBeCalledWith(newQuery);
  });

  it('mutation functions are updated', () => {
    const wrapper = mount(
      // @ts-ignore
      <UrqlClient client={client} query={query} mutations={mutations}>
        {children}
      </UrqlClient>
    );

    const newMutations = {
      newKey: {},
    };

    children.mockClear();
    wrapper.setProps({ mutations: newMutations });

    expect(Object.keys(children.mock.calls[1][0].mutations)).toEqual(
      Object.keys(newMutations)
    );
  });
});

describe('componentWillUnmount', () => {
  it('unsubscribes from client lib', () => {
    const wrapper = shallow(
      // @ts-ignore
      <UrqlClient client={client} query={query}>
        {children}
      </UrqlClient>
    );

    wrapper.instance().componentWillUnmount();
    expect(clientInstance.unsubscribe).toBeCalled();
  });
});

describe('mutation functions', () => {
  it('have same properties as mutation argument', () => {
    const wrapper = shallow(
      // @ts-ignore
      <UrqlClient client={client} query={query} mutations={mutations}>
        {children}
      </UrqlClient>
    );

    expect(Object.keys(children.mock.calls[0][0].mutations)).toEqual(
      Object.keys(mutations)
    );
  });

  it('calls executeMutation', () => {
    const wrapper = shallow(
      // @ts-ignore
      <UrqlClient client={client} query={query} mutations={mutations}>
        {children}
      </UrqlClient>
    );

    const addUser = children.mock.calls[0][0].mutations.addUser;
    addUser();

    expect(clientInstance.executeMutation).toBeCalled();
  });

  it('passes mutation query to executeMutation', () => {
    const wrapper = shallow(
      // @ts-ignore
      <UrqlClient client={client} query={query} mutations={mutations}>
        {children}
      </UrqlClient>
    );

    const removeUser = children.mock.calls[0][0].mutations.removeUser;
    removeUser();

    expect(clientInstance.executeMutation).toBeCalledWith(
      expect.objectContaining({ query: mutations.removeUser.query })
    );
  });

  it('passes mutation vars to executeMutation', () => {
    const wrapper = shallow(
      // @ts-ignore
      <UrqlClient client={client} query={query} mutations={mutations}>
        {children}
      </UrqlClient>
    );

    const removeUser = children.mock.calls[0][0].mutations.removeUser;
    const vars = { name: 'Stefano' };
    removeUser(vars);

    expect(clientInstance.executeMutation).toBeCalledWith(
      expect.objectContaining({ variables: vars })
    );
  });
});

describe('on change from client', () => {
  it('child components are updated', () => {
    const wrapper = shallow(
      // @ts-ignore
      <UrqlClient client={client} query={query} mutations={mutations}>
        {children}
      </UrqlClient>
    );

    const update = {
      fetching: false,
      data: { alert: 'all done' },
      error: { type: 'unimportant' },
    };

    const updateFun = client.createInstance.mock.calls[0][0].onChange;

    children.mockClear();
    updateFun(update);

    expect(children).toBeCalledWith(expect.objectContaining(update));
  });
});
