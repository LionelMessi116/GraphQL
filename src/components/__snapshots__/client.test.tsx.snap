// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on creation defaults passes snapshot 1`] = `
Array [
  Array [
    Object {
      "data": undefined,
      "error": undefined,
      "fetching": true,
      "mutations": Object {
        "addUser": [Function],
        "removeUser": [Function],
      },
      "refetch": [Function],
    },
  ],
]
`;

exports[`passes snapshot 1`] = `
<h1>
  This is a child component
</h1>
`;
